### ISSUE TITLE
Code quality

### ISSUE THREAD

**Issue by pjljvandelaar** [TorXakis/TorXakis#408]
To ensure a high quality of our code
including readability and transferability
our code should have 
* minimal (none?) hlint warnings
* build with --pedantic flag 
* 100% documentation of interface (as measured by haddock)

Note: 
* Coverage versus hlint.
 One needs to use /= to cover "data X deriving (Eq)".
 However, on not (var /= var) hlint generates warning: Use ==
 Is there another way to achieve coverage? Or shouldwe ignore for test files the Use == warning?
* alex issue https://github.com/simonmar/alex/issues/103 prevents a pedanctic build.
* the stack --pedantic flag turns on -Wall -Werror for GHC (all warnings on, and warnings treated as errors)
 * What about hlint : "Suggestion: Use camelCase" when a name contains a underscore, i.e., _? Do we agree with this enforced naming convention?

related to
#87 
#91 
#145

**Comment by pjljvandelaar** [comment]
* alex issue solved
* we already build with pedantic

**Comment by capitanbatata** [comment]
> Do we agree with this enforced naming convention?

Yes! Please.

**Comment by pjljvandelaar** [comment]
We should also measure the coverage of our test cases (and act on it
see #92)

**Comment by pjljvandelaar** [comment]
We should also use weeder to improve code quality: https://github.com/ndmitchell/weeder

**Comment by pjljvandelaar** [comment]
And maybe https://hackage.haskell.org/package/homplexity

**Comment by pjljvandelaar** [comment]
https://www.reddit.com/r/haskell/comments/464wm4/code_quality_metrics_for_haskell/ for more tooling

**Comment by capitanbatata** [comment]
Here are some metrics I got using `argon`. What do we want to do with this? Put it in code climate?

![screenshot from 2018-01-04 14-44-55](https://user-images.githubusercontent.com/175315/34566202-06ad46f4-f15e-11e7-83da-473103159d27.png)

`expand` and `reduce` are really complex (as we probably know...).

**Comment by pjljvandelaar** [comment]
Incorporating it into code climate would be great!

**Comment by pjljvandelaar** [comment]
Coverage should be measured.
For example using [HPC](https://wiki.haskell.org/Haskell_program_coverage)

**Comment by dnadales** [comment]
> Coverage should be measured.

Yes, but we lack tool support. See [this issue](https://github.com/TorXakis/TorXakis/issues/92) for information on what was tried already.
